lib/signs_ui/signs/expiration.ex

```diff
@@ -0,0 +1,76 @@
+defmodule SignsUI.Signs.Expiration do
+  @moduledoc """
+  Handle cleaning up settings in the configuration that have since expired
+  """
+
+  @type state :: %{
+          time_fetcher: (() -> DateTime.t()),
+          loop_ms: integer(),
+          sign_state: pid()
+        }
+
+  @spec start_link(Keyword.t()) :: GenServer.on_start()
+  def start_link(opts \\ []) do
+    name = opts[:name] || __MODULE__
+
+    GenServer.start_link(__MODULE__, opts, name: name)
+  end
+
+  @spec init(Keyword.t()) :: {:ok, state()}
+  def init(opts) do
+    time_fetcher = opts[:time_fetcher] || fn -> DateTime.utc_now() end
+    loop_ms = opts[:loop_ms] || 5_000
+    sign_state = opts[:sign_state] || SignsUI.Signs.State
+    schedule_loop(self(), loop_ms)
+    {:ok, %{time_fetcher: time_fetcher, loop_ms: loop_ms, sign_state: sign_state}}
+  end
+
+  @spec handle_info(:process_expired, state()) :: state()
+  def handle_info(:process_expired, state) do
+    expire_signs(state.time_fetcher, state.sign_state)
+
+    schedule_loop(self(), state.loop_ms)
+
+    {:noreply, state}
+  end
+
+  @spec expire_signs((() -> DateTime.t()), pid()) :: :ok
+  def expire_signs(time_fetcher, sign_state) do
+    signs =
+      SignsUI.Signs.State.get_all(sign_state)
+      |> Enum.map(&expire_single_sign(&1, time_fetcher))
```

Comment:
It might be nice to not update the state (POST to S3, broadcast to listening clients) if we don't have to. What about an approach something like:

```ex
expiration = time_fetcher.()

expired_signs = 
  sign_state_server
  |> SignsUI.Signs.State.get_all()
  |> Enum.flat_map(&expire_sign(&1, expiration)
  |> Enum.into()

if expired_signs != {} do
  SignsUI.State.update_some(sign_state_server, expired_signs)
end

:ok
```

And then `expire_sign/2` would have in it, roughly:

```ex
if expired? do
  [%Sign{...}]
else
  []
end
```

The "`flat_map` as a way to filter+map" approach is one we use fairly frequently here.

Reply:
Good thought, I'll go with that approach.

Reply:
Done